File: analysis/tools/upstream_flow_dfs.R
data <- read.csv('ct_catchments.csv',stringsAsFactors=F);
data <- data[,2:dim(data)[2]]
data[1:10,]
colnames(data) <- c('fid','nid','shape','km');

#[1] Build Adjacency [List-----------
Rev <- vector(mode='list',length=0);
for(i in 1:dim(data)[1]){
  fid<-toString(data[i,'fid']);
  Rev[[fid]] <- c(0);
}
for(i in 1:dim(data)[1]){
  nid<-toString(data[i,'nid']);
  Rev[[nid]] <- c();
}
for(i in 1:dim(data)[1]){
  fid<-toString(data[i,'fid']);
  nid<-toString(data[i,'nid']);
  Rev[[nid]] <- c(Rev[[nid]],fid);
}#----------------------------------

#[2]Traverse the Tree (AL)----------
up_dfs<-function(Ref,i){
  L <- c(i); #visiting node i
  children <- Rev[[i]];
  if(children[1]==0){ #base case
    L                 #stops and returns
  } else {            #general case
    for(j in 1:length(children)){
      L <- c(L,up_dfs(Rev,children[j]));
    }
  }
  L
}#----------------------------------

File: analysis/tools/upstream_flow_dfs.py
import json

#[2] Traverse the Tree---------------------------------------------------
def up_dfs(R,i):
    L = [i]
    if(R[i][0]==0): return L       #base case (no children)
    else:                          #general case
        for j in range(len(R[i])):
            L += [up_dfs(R,j)]     #get the children of children
        return L
#------------------------------------------------------------------------

#[optional] append rows of zero data
def write_upstream_al_json(R,path):
    with open(path,'w') as f:
        json.dump(R,f)
        return True

path = '/home/tbecker/Downloads/ct_catchments.csv'
with open(path,'r') as f:
    raw = [row.replace('\n','').split(',')[1:3] for row in f.readlines()]
data = raw[1:]

#[1] Build Adjacency [List------------------------------------------------
Rev,fids = {},set([])
for i in range(len(data)): #find the parent to child edges
    fids.add(data[i][0])   #keep all fids...
    if data[i][1] in Rev:  Rev[data[i][1]] += [data[i][0]]
    else:                  Rev[data[i][1]]  = [data[i][0]]
for fid in fids.difference(set(Rev)): Rev[fid] = [0] #add terminal
#-------------------------------------------------------------------------
write_upstream_al_json(Rev,'/home/tbecker/Downloads/ct_catchments_upstream_al.json')